### DISCUSSION TITLE
Is it possible to generate bulk update queries (case-when-then) instead of one-by-one update using SQLAlchemy core?

### DISCUSSION THREAD

**Discussion by alvassin** [sqlalchemy/sqlalchemy#6640] [Usage Questions]
I am trying to update huge amount of data in alembic migration using `bindparam`:

```
values = [{"value": ..., "_id": ...}]
query = example_table.update().values({
    'value': bindparam('value')
}).where(
    example_table.columns.id == bindparam('_id')
)
conn.execute(query, values)
```

And i found in wireshark sqlalchemy emit single query for each array element, so it takes a lot of time:

```
UPDATE example SET value = "value1" WHERE id = 1;
```

Is it possible to make single query like following (should run much faster)?

```
UPDATE example SET value = CASE
   WHEN id = 1 THEN "value1"
   WHEN id = 2 THEN "value2"
END
WHERE id IN (1, 2)
```

I found examples for [select + case](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.case), but i can't get how to make it work with updates.

**Comment by CaselIT** [comment]
Hi,

This is really deoends on the dbapi you are using. What db and dbapi was that observed?

To see what parameters are passed by sqlalchemy to the dbapi, you can set `echo=True` in the create engine call

**Reply by alvassin** [reply]
Thank you for response, I am using postgres with default psycopg2 driver:

```
from sqlalchemy import Column, MetaData, Table, Integer, create_engine, bindparam

meta = MetaData()
example = Table(
    'example',
    meta,
    Column('id', Integer, primary_key=True),
    Column('value', Integer),
)

engine = create_engine("postgresql+psycopg2://user:hackme@localhost:5432/mydb")
meta.drop_all(engine)
meta.create_all(engine)
with engine.begin() as conn:
    query = example.insert().values([{'id': i, 'value': i} for i in range(100)])
    conn.execute(query)

engine.echo = True
query = example.update().values({'value': bindparam('value')}).where(example.c.id == bindparam('id_'))
with engine.begin() as conn:
    conn.execute(query, [{'id_': i, 'value': i*2} for i in range(100)])


# UPDATE example SET value=%(value)s WHERE example.id = %(id_)s
# is executed multiple times, one query per list item

# I need to make one big case-when-then request instead of 100 small requests
```

**Reply by CaselIT** [reply]
As you can see sqlalchemy passes the statement in the execute many form.

```
2021-06-15 23:04:13,784 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-06-15 23:04:13,786 INFO sqlalchemy.engine.Engine UPDATE example SET value=%(value)s WHERE example.id = %(id_)s
2021-06-15 23:04:13,787 INFO sqlalchemy.engine.Engine [generated in 0.00076s] ({'value': 0, 'id_': 0}, {'value': 2, 'id_': 1}, {'value': 4, 'id_': 2}, {'value': 6, 'id_': 3}, {'value': 8, 'id_': 4}, {'value': 10, 'id_': 5}, {'value': 12, 'id_': 6}, {'value': 14, 'id_': 7}  ... displaying 10 of 100 total bound parameter sets ...  {'value': 196, 'id_': 98}, {'value': 198, 'id_': 99})
2021-06-15 23:04:13,892 INFO sqlalchemy.engine.Engine COMMIT
```

That said, I think the issue is on psycopg2 side, that seems to not use the execute many form. I would suggest looking at it's documentation and https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.psycopg2 to see if some configuration helps

Regarding your original question, I think is should be possible to do that, but there is no automatic utility in sqlalchemy do to that, so it needs to be done manually. Something like this
```py
with engine.begin() as conn:
    values = [{"id_": i, "value": i * 2} for i in range(100)]
    query = (
        example.update()
        .values({"value": case(*((example.c.id == v["id_"], v["value"]) for v in values))})
        .where(example.c.id.in_([v["id_"] for v in values]))
    )
    conn.execute(query)
```

Note that there may be a limit at how many parameters you can pass in the query (iirc it's something like 30k)

**Reply by zzzeek** [reply]
psycopg2 sends the statement multiple times unless the "fast execution helpers" are used.  these default to off for UPDATE/DELETE statements in SQLAlchemy's psycopg2 dialect.

the case-when-then version can be implemented using [case](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=case#sqlalchemy.sql.expression.case) - to use this with UPDATE just use the construct in the `where()` method.

**Reply by alvassin** [reply]
Thank you very much for examples, worked like a charm!

btw, @CaselIT, there should be no parameter limits in that case, as psycopg2 uses simple query protocol and only client-side escaping. As i remember there is only limit for extended query parameters (about ~32k) in PostgreSQL protocol. 

Limit would is actual with asyncpg driver (there is problem with huge inserts using asyncpgsa, for example, but there is no such with SQLAlchemy, as it is doing simple statement and Bind-Execute combination for each row then).
